### DISCUSSION TITLE
hibernate_sequence considered harmful

### DISCUSSION THREAD

**Discussion by gavinking** [hibernate/hibernate-orm#3809] [Design Proposals]
On databases which prefer sequence id generation, `GenerationType.AUTO` defaults to use of a single sequence with the name `hibernate_sequence`. 

This is bad for two reasons:

1. pretty sure nobody actually wants a sequence named `hibernate_sequence` in their database, and
2. it means that the behavior on a platform which prefers sequences is quite different to the behavior on a platform which prefers identity columns: in one case the ids are unique at the table level, and in the other they're unique at the PU level.

Now, I need to do some more reading, but IMO, `GenerationType.AUTO` doesn't have any really well-defined semantics in JPA, and we can legitimately change its behavior to create a sequence based on the corresponding table name.

And I think we *should* do this in H6.

**Comment by sebersole** [comment]
To be clear... you are talking about just the name, right?  

I bring this up because on the source issue in Quarkus there seemed to be some suggestions that we change the strategy that AUTO maps to, which I completely disagree with.

What you want is to change the strategy for changing the implicit name of the database structure backing the SequenceStyleGenerator, similar to `SequenceStyleGenerator#CONFIG_PREFER_SEQUENCE_PER_ENTITY` and `SequenceStyleGenerator#CONFIG_SEQUENCE_PER_ENTITY_SUFFIX`.  You just want a different choice here.

What I would suggest is this...  Take these settings I mentioned and extract them into a strategy (`ImplicitGeneratorNamingStrategy` ?) and provide the ability to plug-in alternatives.  In fact I'd even go so far as to create short-named versions for a few different naming strategies.

Or maybe even a method on `ImplicitNamingStrategy` once we get that all hooked in properly.  As I mentioned on the Quarkus issue, the big thing there is what to base the name on consistently considering that we have to account for entities as well as id-bags.

You sometimes like to scoff at backwards compatibility, and I am uber jealous that we cannot do that in ORM; but we can't - we have to consider backwards compatibility.  We all agree the name sucks, but it has unfortunately sucked for the last 20 years.  Whatever choice we make here needs to not break that.

Actually I think the existing settings may even be fine for your needs.  The problem with extending that to your needs is that atm they are specific to each generator.  We could look at a cascade setting: local > global.  But that would give Quarkus, Panache and users the ability to control that for the whole SF (aside from local overrides).

So a few nice options I think.

And yes, I think we should finally change this in 6

**Reply by gavinking** [reply]
I'm not talking about changing the id generation strategy from sequence to something else like table. I'm just saying there should be one sequence per `@Entity` hierarchy.

So if I have an entity mapped to the table `BOOK`, then by default its sequence would be called `BOOK_IDS` or whatever.

And those sequences would be automatically created for me.

**Reply by gavinking** [reply]
> You sometimes like to scoff at backwards compatibility

I don't scoff at it, I just say: "we should own up to our mistakes and fix them eventually, instead of leaving things broken _forever_".

It's not that backwards compatibility doesn't matter; of course it matters. But the whole point of doing a new major version is that we get to break stuff. Now, of course we shouldn't just _arbitrarily_ break stuff for no good reason, but when we've made a mistake, this is our chance to make amends for that.

And what really matters with respect to compatibility is that there's an easy migration path for people who're depending on the old behaviour, and in this case it could be something as simple as a config property.

**Reply by gavinking** [reply]
> Or maybe even a method on `ImplicitNamingStrategy` once we get that all hooked in properly.

Yes, I was already thinking that the mapping from table name -> sequence name would be handled by a `NamingStrategy`.

**Reply by gavinking** [reply]
> similar to `SequenceStyleGenerator#CONFIG_PREFER_SEQUENCE_PER_ENTITY` and `SequenceStyleGenerator#CONFIG_SEQUENCE_PER_ENTITY_SUFFIX`.

Yes, I think those are the behaviours I'm looking for. Really I just think that `SequenceStyleGenerator#CONFIG_PREFER_SEQUENCE_PER_ENTITY` has the wrong default.

**Reply by sebersole** [reply]
> I'm not talking about changing the id generation strategy from sequence to something else like table. I'm just saying there should be one sequence per `@Entity` hierarchy.
> 
> So if I have an entity mapped to the table `BOOK`, then by default its sequence would be called `BOOK_IDS` or whatever.
> 
> And those sequences would be automatically created for me.

This is essentially what happens with those settings I mentioned.  You just get the entity name used instead of the table name.  Meh.  Not that big of a deal to me.  Would it ideally be the table name?  Meh.  Sure, ok, maybe.  But its 50/50, especially considering the original requests asked for entity name as the base.

> > You sometimes like to scoff at backwards compatibility
> 
> I don't scoff at it, I just say: "we should own up to our mistakes and fix them eventually, instead of leaving things broken _forever_".
> 
> It's not that backwards compatibility doesn't matter; of course it matters. But the whole point of doing a new major version is that we get to break stuff. Now, of course we shouldn't just _arbitrarily_ break stuff for no good reason, but when we've made a mistake, this is our chance to make amends for that.
> 
> And what really matters with respect to compatibility is that there's an easy migration path for people who're depending on the old behaviour, and in this case it could be something as simple as a config property.

I totally agree changing things is totally fine in a major.  So much so I often get flak for changing too much because I believe this also.  Schemas are different.  There I take super extra care to minimize changes.

Could the user manually change the schema to meet our new expectations?  Sure.  Is that an "easy migration path"?  Well YMMV, but I know I'd hate having to alter my schema just for a new ORM release.


> > Or maybe even a method on `ImplicitNamingStrategy` once we get that all hooked in properly.
> 
> Yes, I was already thinking that the mapping from table name -> sequence name would be handled by a `NamingStrategy`.

I have broken out NamingStrategy into 2 different contracts because cramming it all into one is a terrible idea and leads to the mess we have now with naming-strategies.  But, sure I could see a method on ImplicitNamingStrategy to handle implicit sequence names.  But, that's actually not even what you are looking for.  You want this to happen in the specific way you want automatically.  That is where I start to have difference of opinion.

> > similar to `SequenceStyleGenerator#CONFIG_PREFER_SEQUENCE_PER_ENTITY` and `SequenceStyleGenerator#CONFIG_SEQUENCE_PER_ENTITY_SUFFIX`.
> 
> Yes, I think those are the behaviours I'm looking for. Really I just think that `SequenceStyleGenerator#CONFIG_PREFER_SEQUENCE_PER_ENTITY` has the wrong default.

I disagree it has the wrong default.  You keep missing the point of backwards compatibility.  The default here was chosen to maintain backwards compatibility.  It's just the "wrong" default for you because it does not do what you want.

**Reply by sebersole** [reply]
Look, in general I agree a different name for the sequence would be better.  We can argue whether a single sequence or a per-entity sequence is better.  Either way you have a means to get that behavior.   What I think is missing (and think you agree) is a way to globally change that behavior.

**Reply by gavinking** [reply]
> You just get the entity name used instead of the table name.

Entity name is fine.

>  I often get flak for changing too much because I believe this also.

Ok, but this is just the difference between a delta function of flack centered on the major release date, and the integral of much lower-intensity flak over time. :-)

> You want this to happen in the specific way you want automatically.

I just think the _default behaviour_ should be that `AUTO` means ids are unique only across a hierarchy, no matter what the platform's preferred way of generating ids is. Because otherwise there is a weird difference in default semantics across platforms.

>  It's just the "wrong" default for you because it does not do what you want.

Right. It's "wrong" because if I were designing this from scratch knowing what I know now, it's not the behavior I would pick.

That's always what I mean by "wrong" :-)

**Reply by sebersole** [reply]
> > I often get flak for changing too much because I believe this also.
> 
> Ok, but this is just the difference between a delta function of flack centered on the major release date, and the integral of much lower-intensity flak over time. :-)

Haha.  You speak the true true.

 
> > You want this to happen in the specific way you want automatically.
> 
> I just think the _default behaviour_ should be that `AUTO` means ids are unique only across a hierarchy, no matter what the platform's preferred way of generating ids is. Because otherwise there is a weird difference in default semantics across platforms.

I do not follow this point.  For AUTO I pick `SequenceStyleGenerator` which uses a sequence or a table depending on the db's capabilities.  There is no difference of semantic.  Unless I miss what you mean?


> > It's just the "wrong" default for you because it does not do what you want.
> 
> Right. It's "wrong" because if I were designing this from scratch knowing what I know now, it's not the behavior I would pick.
> 
> That's always what I mean by "wrong" :-)

Fair enough.  Me too.

**Reply by gavinking** [reply]
> For `AUTO` I pick `SequenceStyleGenerator` which uses a sequence or a table depending on the db's capabilities.

Oh, it never picks auto increment/identity columns? I thought it could.

**Reply by sebersole** [reply]
No, thats the old native stuff which did in fact have the difference in semantic you mention.  This was added to fix that

**Reply by gavinking** [reply]
Ah, OK, I stand corrected.

**Reply by sebersole** [reply]
People (ok Vlad) have suggested that picking IDENTITY would be more performant.  But I seriously doubt that with allocation-size,  And he never gave us benchmark nor numbers.

So I think this SequenceStyleGenerator solution is the best solution

**Reply by gavinking** [reply]
I don't think I agree with Vlad. I think the "consistent behavior" argument is stronger than the (fairly unconvincing, it seems to me) performance argument.

**Reply by gavinking** [reply]
Question about that though: by default does `AUTO` use block allocation?

I guess I think it should.

**Reply by sebersole** [reply]
Good call.  I thought it did already, but added a test and that is not what is happening currently.  I don't think it is a great idea to make that change for 5.x and would add that to my vote proposal below for 6.x

**Reply by sebersole** [reply]
For posterity sake, the change in 5.x would be in `org.hibernate.cfg.BinderHelper#makeIdGenerator`, specifically the block:

```
if ( !isEmptyAnnotationValue( generatorName ) ) {
    ...
}
```

ATM there is no `else` block, but we'd simply add:

```
if ( !isEmptyAnnotationValue( generatorName ) ) {
    ...
}
else {
    // NOTE: we choose the constant from `SequenceStyleGenerator` but the parameter
    // key is actually the same for all generators that understand an allocation size
    params.setProperty( SequenceStyleGenerator.INCREMENT_PARAM, "50" );
}
```

**Reply by sebersole** [reply]
Workaround btw is to map this as:

```
@Id
@GeneratedValue( strategy = AUTO, generator = "sequence" )
@SequenceGenerator( name = "sequence" )
```

Not great to have to add the `@SequenceGenerator` but meh

**Reply by sebersole** [reply]
> > You just get the entity name used instead of the table name.
> 
> Entity name is fine.

Actually looking at that code for something else, we do in fact have the table name(s) available.  This would be an even bigger user-facing change but we could decide to use the table name instead.  In retrospect that would have been a better option

**Comment by gavinking** [comment]
So it seems the situation here is better than I thought. I had thought that `AUTO` sometimes meant autoincrement, but apparently that's not the case. And indeed, it seems a *good* thing that it's not the case. At least the behavior is pretty consistent across platforms.

Nevertheless, after sleeping on this, I still think that a sequence-per-hierarchy would be preferable to a single global sequence with a weird name. And I would still prefer to change it to work like that (by default) in H6.

WDYT, @sebersole?

**Reply by sebersole** [reply]
In isolation I totally agree that sequence-per-hierarchy (aka the existing "prefer sequence for entity" handling) would be the right choice.

I have the same concern I always have when we change things related to what we expect the user's relational schema to look like for the same config.  To a large degree I can get behind this one though.

So personally I am fine with that.  We can mention it in the migration guide.

Something we discussed today was a way to mark these discussions as "resolved" and going ahead with work on it.  What we came up with in the absence of a proper "resolve" feature from GH is to vote to see if we have a consensus. 

So let's call a vote on this, specifically the part about preferring sequence-per-hierarchy for AUTO.  

There is a difficulty here though... atm (5.x) much of this is mucked up by OGM requirements.  IMO there is a better solution for OGM and, TBH, I think we should consider stopping adding funky support for OGM and actually actively removing things like this that stand as a road-block to doing something better in ORM

All that said, I propose that we change the behavior to prefer preferring sequence-per-entity.

Additionally I propose that we clean up the fugliness here coming from the OGM "requirment", specifically:

1) drop `IdGeneratorStrategyInterpreter` (which was added for OGM) and simply expect that OGM would register a custom resolution with `IdentifierGeneratorFactory`
2) change the internal `IdentifierGeneratorFactory#generatorStrategyToClassNameMap` field from `Map<String,Class>` to some form of Collection of functional contract references based on `GenerationType`, `IdentifierGeneratorDefinition`, etc
3) change the contract of `IdentifierGeneratorFactory#createIdentifierGenerator` to pass along additional details like `GenerationType` and `IdentifierGeneratorDefinition` or pass them in the Map passed to `#createIdentifierGenerator` (though personally I very much dislike the "Map as an API" paradigm) 

These additional (numbered) proposed changes simply make it easier to react to AUTO.  As it stands atm, by the time we try to create the `IdentifierGenerator` the actual `GenerationType` is no longer available.

+1 / -1 ?

P.S. Per https://github.com/hibernate/hibernate-orm/discussions/3809#discussioncomment-464888 I also think we should add the JPA default allocation/increment size of 50 when we have `GenerationType#AUTO`.  This is a case where the proposed make this a lot easier/cleaner to achieve

**Comment by sebersole** [comment]
See:

* https://hibernate.atlassian.net/browse/HHH-14491
* https://hibernate.atlassian.net/browse/HHH-14492
* https://hibernate.atlassian.net/browse/HHH-14495
* https://hibernate.atlassian.net/browse/HHH-14496
